stan/math/rev/fun/elt_multiply.hpp

```diff
@@ -33,13 +34,10 @@ auto elt_multiply(const Mat1& m1, const Mat2& m2) {
     arena_t<promote_scalar_t<var, Mat2>> arena_m2 = m2;
     arena_t<ret_type> ret(arena_m1.val().cwiseProduct(arena_m2.val()));
     reverse_pass_callback([ret, arena_m1, arena_m2]() mutable {
-      for (Eigen::Index j = 0; j < arena_m2.cols(); ++j) {
-        for (Eigen::Index i = 0; i < arena_m2.rows(); ++i) {
-          const auto ret_adj = ret.adj().coeffRef(i, j);
-          arena_m1.adj().coeffRef(i, j) += arena_m2.val().coeff(i, j) * ret_adj;
-          arena_m2.adj().coeffRef(i, j) += arena_m1.val().coeff(i, j) * ret_adj;
-        }
-      }
+      auto ret_adj = ret.adj().array();
+      eigen_results(arena_m1.adj().array(), arena_m2.adj().array())
+          += eigen_expressions(arena_m2.val().array() * ret_adj,
+                               arena_m1.val().array() * ret_adj);
```

Comment:
Should we do a sanity check in `eigen_results` to make sure there are not expressions that get passed to `eigen_results`

Reply:
Also I'm still trying to think about how to stop users of this from calling something like `arena_m2.val() * arena_m2.val()`. 

I'm playing around with the below, is there anything else we would not want to have in `eigen_expressions()` besides products? decompositions? Do you think there is some way we can detect whether an expression has a `Eigen::Product<>` in it's template parameters? The only thing I could think of was some recursive type trait that would keep peeking through the template params. I just really worry that can become a bad footgun if we are not paying attention

https://godbolt.org/z/efTfjY

Reply:
> how to stop users of this from calling something like arena_m2.val() * arena_m2.val().

If it's not easy to do something like this, I'm okay leaning on review + benchmarks (when we get them) to catch these things. Always a chance we accidentally code up really slow stuff that passes tests.

Reply:
Yeah it would be nice to at least catch the simpler cases. @t4c1 if you spend 20 mins goofing around to find a nice solution to this and can't make one then I'm fine with leaving it as is

Reply:
I changed this to use Eigen's evaluators. This way we efficiently handle expressions that are better off being evaluated separately (such as matrix product). In most cases this should be no better than evaluating such expressions separately, but it should also not be slower.